### PR TITLE
A gdb user defined command to print kernel memory pools statistics.

### PR DESCRIPTION
**Pull request by pwit81**
This is a simple python script implementing gdb command of the form `kmem pool`, where pool is one of pools defined by MALLOC_DEFINE, e.g. M_TEMP, M_VFS, M_FD.  Invoke gdb command `help kmem` for details.

### CODE REVIEW

**Comment by cahirwpz** [comment]
Ok, I'm back into the game after a long break.

The printout from `kmem` command is not very informative and cannot replace `kmem_dump` at the moment. To be more precise, when I entered gdb command `call kmem_dump(M_VFS)` at randomly chosen moment I got following:

```
[kmem] pool at 0x80121830:
[kmem] > malloc_arena 0x80e28020 - 0x80e29000:
[kmem]    U 0x80e28020 32
[kmem]    U 0x80e28050 8
[kmem]    U 0x80e28068 32
[kmem]    U 0x80e28098 8
[kmem]    U 0x80e280b0 32
[kmem]    U 0x80e280e0 8
[kmem]    U 0x80e280f8 56
[kmem]    U 0x80e28140 56
[kmem]    U 0x80e28188 32
[kmem]    U 0x80e281b8 8
[kmem]    U 0x80e281d0 32
[kmem]    U 0x80e28200 8
[kmem]    U 0x80e28218 32
[kmem]    U 0x80e28248 8
[kmem]    U 0x80e28260 32
[kmem]    U 0x80e28290 16
[kmem]    U 0x80e282b0 32
[kmem]    U 0x80e282e0 8
[kmem]    U 0x80e282f8 32
[kmem]    U 0x80e28328 16
[kmem]    F 0x80e28348 3240
```

General format of output should be:
```
pool information:
> arena[0] information
 > arena[0].block[0] information
 > arena[0].block[1] information
 ...
> arena[1] information
 ...
```

Please note that the output must contain used and free blocks.

File: debug/kmem.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        self.pool_names = Kmem.__global_pool_names()
+
+    @staticmethod
+    def __is_local_pool_name(var):
+        return vars.has_type(var, 'kmem_pool_t *')
+
+    @staticmethod
+    def __is_global_pool_name(var):
+        return vars.has_type(var, 'kmem_pool_t *') or \
+            vars.has_type(var, 'static kmem_pool_t *')
```
Comment (cahirwpz) [inline debug/kmem.py]: I made a quick experiment with gdb:
```
(gdb) break malloc_multiple_allocations
(gdb) n
(gdb) n
(gdb) n
(gdb) python print(gdb.parse_and_eval('M_VMOBJ').type)
kmem_pool_t *
(gdb) python print(gdb.parse_and_eval('M_TEMP').type)
kmem_pool_t *
(gdb) python print(gdb.parse_and_eval('mp').type)
kmem_pool_t *
```

`M_VMOBJ` is defined globally as static variable, `M_TEMP` is defined globally without static qualifier and `mp` is defined on stack. All three of them have the same type. Have you ever triggered a situation when gdb reports type of a variable as `static kmem_pool_t *` ?

File: debug/utils.py [hunk 2]
```diff
@@ -56,6 +56,10 @@ def dump(self):
     def display_hint(self):
         return 'map'
 
+    @property
+    def address(self):
+        return str(self._obj.address)
```
Comment (cahirwpz) [inline debug/utils.py]: Natural type for `self._obj.address` is integer. Please do not cast it to string by default.

File: debug/kmem.py (hunk 1 above)
Comment (pwit81) [reply debug/kmem.py]: No. Distinction between static and non-static types was relevant in earlier commits, i.e. when pool names were read from gdb using `info variables`.  Corrected in cf3bf8d.

File: debug/utils.py (hunk 2 above)
Comment (pwit81) [reply debug/utils.py]: Corrected in cf3bf8d.

**Comment by cahirwpz** [comment]
@pwit81 This task is not yet finished. Look at my previous comment. Printout from `kmem` command still does not report used blocks.

**Comment by cahirwpz** [comment]
I decided to close this PR since `kmalloc` memory manager is being completely rewritten.

**Pull request closed without merging**